python/lsst/dax/ppdb/sql/_base.py

```diff
@@ -0,0 +1,475 @@
+        else:
+            # Missing schema version is identical to 0.1.0
+            version = VersionTuple(0, 1, 0)
+
+        metadata = sqlalchemy.schema.MetaData(schema=schema_name)
+
+        converter = ModelToSql(metadata=metadata)
+        converter.make_tables(schema.tables)
+
+        return metadata, version
+
+    @classmethod
+    def create_replica_chunk_table(cls, table_name: str | None = None) -> schema_model.Table:
+        """Create the ``PpdbReplicaChunk`` table which will be added to the
+        schema.
+
+        Parameters
+        ----------
+        table_name : `str` or `None`
+            Name of the table to create. If not provided, defaults to
+            "PpdbReplicaChunk".
+        """
+        table_name = table_name or "PpdbReplicaChunk"
+        columns = [
+            schema_model.Column(
+                name="apdb_replica_chunk",
+                id=f"#{table_name}.apdb_replica_chunk",
+                datatype=felis.datamodel.DataType.long,
+            ),
+            schema_model.Column(
+                name="last_update_time",
+                id=f"#{table_name}.last_update_time",
+                datatype=felis.datamodel.DataType.timestamp,
+                nullable=False,
+            ),
+            schema_model.Column(
+                name="unique_id",
+                id=f"#{table_name}.unique_id",
+                datatype=schema_model.ExtraDataTypes.UUID,
+                nullable=False,
+            ),
+            schema_model.Column(
+                name="replica_time",
+                id=f"#{table_name}.replica_time",
+                datatype=felis.datamodel.DataType.timestamp,
+                nullable=False,
+            ),
+        ]
+        indices = [
+            schema_model.Index(
+                name="PpdbInsertId_idx_last_update_time",
+                id="#PpdbInsertId_idx_last_update_time",
+                columns=[columns[1]],
+            ),
+            schema_model.Index(
+                name="PpdbInsertId_idx_replica_time",
+                id="#PpdbInsertId_idx_replica_time",
+                columns=[columns[3]],
+            ),
+        ]
+
+        # Add table for replication support.
+        chunks_table = schema_model.Table(
+            name=table_name,
+            id=f"#{table_name}",
+            columns=columns,
+            primary_key=[columns[0]],
+            indexes=indices,
+            constraints=[],
+        )
+        return chunks_table
+
+    def get_table(self, name: str) -> sqlalchemy.schema.Table:
+        for table in self._sa_metadata.tables.values():
+            if table.name == name:
+                return table
+        raise LookupError(f"Unknown table {name}")
+
+    @property
+    def schema_version(self) -> VersionTuple:
+        """Version of the APDB database schema (`VersionTuple`)."""
+        return self._schema_version
+
+    @classmethod
+    def filter_tables(cls, schema_dict: dict[str, Any]) -> list[Any]:
```

Comment:
I'm not sure that you need the whole schema dictionary here, can you just pass the list of tables? And change `list[Any]` to `list[dict]`? Or maybe replace this method with the method that filters the names of the tables? My concern is that this method needs to know about felis schema ("tables" key or "name" key) which is potentially fragile.

Reply:
What if it returned a `list[Table]` using the `schema_model` types?

Then it wouldn't be dealing with the raw dictionary data.

Reply:
But the only use for this not is to filter tables based on their names, you do not need `Table` for that.

Reply:
I think that would be nice to have a standalone method for making the `metadata` table independently from the YAML schema. I can work on that though it would be duplicative.

- I was kind of wondering why the metadata table is in the schema at all though since it isn't really user-facing. Wouldn't make sense for the APDB to be able to create it without needing the schema? (Either way, I agree it is overkill to read the whole schema just for that table definition.)

Reply:
I changed this so that the sub-class returns a `str` collection of table names. Ones not in this list are then filtered out of the dictionary before creating the schema.